src/credentials_rotators/npm-token-rotation/src/stacks/base-stack.ts

```diff
@@ -0,0 +1,117 @@
+import * as core from "@aws-cdk/core";
+import * as lambda from "@aws-cdk/aws-lambda";
+import * as iam from "@aws-cdk/aws-iam";
+import * as secretsManager from "@aws-cdk/aws-secretsmanager";
+import * as sns from "@aws-cdk/aws-sns";
+import * as snsSubscriptions from "@aws-cdk/aws-sns-subscriptions";
+import { SecretDetail, AccessTokenRotationConfig } from "./types";
+import { StateMachine } from "@aws-cdk/aws-stepfunctions";
+import { SnsAction } from "@aws-cdk/aws-cloudwatch-actions";
+
+type BaseStackProps = core.StackProps & {};
+
+export class BaseStack extends core.Stack {
+  constructor(scope: core.Construct, id: string, props: core.StackProps = {}) {
+    super(scope, id, props);
+  }
+
+  protected grantSecretsManagerToAccessLambda = (fn: lambda.IFunction) => {
+    const principal = new iam.ServicePrincipal("secretsmanager.amazonaws.com");
+    fn.addPermission("invoke_access_to_secretsmanager", { principal });
+  };
+
+  protected grantLambdaFunctionToAccessStepFunctions = (
+    fn: lambda.IFunction,
+    machine: StateMachine
+  ) => {
+    machine.grantStartExecution(fn);
+  };
+
+  protected grantLambdaAccessToRotateSecrets = (
+    fn: lambda.IFunction,
+    secret: SecretDetail
+  ) => {
+    const actions = secret.roleArn
+      ? ["sts:AssumeRole"]
+      : [
+          "secretsmanager:DescribeSecret",
+          "secretsmanager:GetSecretValue",
+          "secretsmanager:PutSecretValue",
+          "secretsmanager:UpdateSecretVersionStage",
+        ];
+    const resources = secret.roleArn ? [secret.roleArn] : [secret.arn];
+
+    fn.addToRolePolicy(
+      new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        resources,
+        actions,
+      })
+    );
+
+    fn.addToRolePolicy(
+      new iam.PolicyStatement({
+        effect: iam.Effect.ALLOW,
+        resources: ["*"],
+        actions: ["secretsmanager:GetRandomPassword"],
+      })
+    );
+  };
+
+  protected grantLambdaAccessToSecrets = (
```

Comment:
Isn't this covered in grantLambdaAccessToRotateSecrets?

Reply:
This only grants  read access and not write access like in the case of rotation. The permission granted here is `secretsmanager:GetSecretValue` where as in case of rotation we grant `secretsmanager:DescribeSecret`,`secretsmanager:GetSecretValue`,`secretsmanager:PutSecretValue` and `secretsmanager:UpdateSecretVersionStage`. When the secret is stored in a different account and access using assume role, the permissions are controlled by the role policy.